rafx-api/src/backends/gles2/internal/gl_context/web.rs

```diff
@@ -0,0 +1,1124 @@
+use crate::gles2::gles2_bindings::types::*;
+use crate::gles2::{
+    gles2_bindings, ActiveUniformInfo, BufferId, FramebufferId, ProgramId, RenderbufferId,
+    ShaderId, TextureId, WindowHash, NONE_BUFFER, NONE_FRAMEBUFFER, NONE_PROGRAM,
+    NONE_RENDERBUFFER, NONE_TEXTURE,
+};
+use crate::{RafxError, RafxResult};
+use fnv::FnvHashMap;
+use raw_window_handle::HasRawWindowHandle;
+use std::ffi::{CStr, CString};
+use std::sync::atomic::Ordering;
+use std::sync::Mutex;
+use wasm_bindgen::JsValue;
+use web_sys::{
+    WebGlBuffer, WebGlFramebuffer, WebGlProgram, WebGlRenderbuffer, WebGlRenderingContext,
+    WebGlShader, WebGlTexture, WebGlUniformLocation,
+};
+
+pub struct GetActiveUniformMaxNameLengthHint;
+
+//TODO: Refactor to avoid integer IDs/hash map lookups
+static NEXT_GL_TEXTURE_ID: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(1);
+static NEXT_GL_BUFFER_ID: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(1);
+static NEXT_GL_SHADER_ID: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(1);
+static NEXT_GL_FRAMEBUFFER_ID: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(1);
+//static NEXT_GL_PROGRAM_ID: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(1);
+//static NEXT_GL_RENDERBUFFER_ID: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(1);
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct LocationId(WebGlUniformLocation);
+
+fn convert_js_to_i32(value: &JsValue) -> Option<i32> {
+    if let Some(value) = value.as_f64() {
+        Some(value as i32)
+    } else if let Some(value) = value.as_bool() {
+        if value {
+            Some(1)
+        } else {
+            Some(0)
+        }
+    } else {
+        None
+    }
+}
+
+pub struct GlContext {
+    context: WebGlRenderingContext,
+    window_hash: WindowHash,
+    textures: Mutex<FnvHashMap<TextureId, WebGlTexture>>,
+    buffers: Mutex<FnvHashMap<BufferId, WebGlBuffer>>,
+    shaders: Mutex<FnvHashMap<ShaderId, WebGlShader>>,
+    programs: Mutex<FnvHashMap<ProgramId, WebGlProgram>>,
+    renderbuffers: Mutex<FnvHashMap<RenderbufferId, WebGlRenderbuffer>>,
+    framebuffers: Mutex<FnvHashMap<FramebufferId, WebGlFramebuffer>>,
+}
+
+impl PartialEq for GlContext {
+    fn eq(
+        &self,
+        other: &Self,
+    ) -> bool {
+        self.window_hash == other.window_hash
+    }
+}
+
+impl GlContext {
```

Comment:
Part of me wonders if this + the native context should be implementing a common trait.

Reply:
I see the reasoning, but I think it's just another thing to maintain that doesn't really do anything. It's not possible to compile them both into the same binary due to them targeting different platforms.